ext/semian/sliding_window.c

```diff
@@ -164,27 +196,33 @@ VALUE semian_simple_sliding_window_reject(VALUE self) {
 
   semian_simple_sliding_window_t *res;
   TypedData_Get_Struct(self, semian_simple_sliding_window_t, &semian_simple_sliding_window_type, res);
-  semian_simple_sliding_window_shared_t *window = get_window(res->key);
 
-  // Store these values because we're going to be modifying the buffer.
-  int start = window->start;
-  int length = window->length;
-
-  int cleared = 0;
-  for (int i = 0; i < length; ++i) {
-    int index = (start + i) % length;
-    int value = window->data[index];
-    dprintf("  i:%d index: %d value:%d max_size:%d length:%d start:%d end:%d", i, index, value, window->max_size, window->length, window->start, window->end);
-    VALUE y = rb_yield(RB_INT2NUM(value));
-    if (RTEST(y)) {
-      if (cleared++ != i) {
-        rb_raise(rb_eArgError, "reject! must delete monotonically");
+  sem_meta_lock(res->sem_id);
+  {
+    semian_simple_sliding_window_shared_t *window = get_window(res->key);
+
+    // Store these values because we're going to be modifying the buffer.
+    int start = window->start;
+    int length = window->length;
+
+    int cleared = 0;
+    for (int i = 0; i < length; ++i) {
+      int index = (start + i) % length;
+      int value = window->data[index];
+      dprintf("  i:%d index: %d value:%d max_size:%d length:%d start:%d end:%d", i, index, value, window->max_size, window->length, window->start, window->end);
+      VALUE y = rb_yield(RB_INT2NUM(value));
+      if (RTEST(y)) {
+        if (cleared++ != i) {
+          sem_meta_unlock(res->sem_id);
```

Comment:
@michaelkipper It took me a while to figure this one out, is this an ok place to release the lock in the event of an exception? `rb_raise` means that the loop exits so `sem_meta_unlock` doesn't get called unless it's here.

Reply:
If it happens enough, it might be worth defining:
```c
void raise_arg_error(uint64_t sem_id, msg, ...) {
  if (sem_id) sem_meta_unlock(sem_id);
  rb_raise(rb_eArgError, msg, __VA_ARGS__); // <-- can't remember the syntax of this...
}
```
and then calling it like:
```c
if (RTEST(y)) {
  if (cleared++ != i) {
    raise_arg_error(res->sem_id, "something...");
  }
}
```